src/modules/flow/http-client/http-client.c

```diff
@@ -1384,4 +1395,286 @@ get_key_process(struct sol_flow_node *node, void *data,
     return 0;
 }
 
+static void
+create_url_close(struct sol_flow_node *node, void *data)
+{
+    struct create_url_data *mdata = data;
+
+    free(mdata->protocol);
+    free(mdata->server);
+    free(mdata->path);
+    free(mdata->fragment);
+    sol_http_param_free(&mdata->params);
+    sol_arena_del(mdata->arena);
+}
+
+static int
+add_query(struct sol_arena *arena, struct sol_http_param *params,
+    const struct sol_str_slice key, const struct sol_str_slice value)
+{
+    char *cvalue, *ckey;
+
+    ckey = sol_arena_strdup_slice(arena, key);
+    SOL_NULL_CHECK(ckey, -ENOMEM);
+
+    cvalue = sol_arena_strdup_slice(arena, value);
+    SOL_NULL_CHECK(cvalue, -ENOMEM);
+
+    if (!sol_http_param_add(params,
+        SOL_HTTP_REQUEST_PARAM_QUERY(ckey, cvalue))) {
+        SOL_ERR("Could not add the HTTP param %s:%s", ckey, cvalue);
+        return -ENOMEM;
+    }
```

Comment:
hey, I'm seeing this pattern of arena + param_add in many places: node-type, tests, internal to sol-http. Please make a helper structure that does it in one place:

```c
struct sol_http_param {
#define SOL_HTTP_PARAM_API_VERSION (1)
    uint16_t api_version;
    uint16_t reserved;

    struct sol_vector params;
    struct sol_arena *arena; // <--- addition
};

SOL_API bool
sol_http_param_add_copy(struct sol_http_param *params, struct sol_http_param_value value)
{
    struct sol_http_param_value *ptr;

    SOL_NULL_CHECK(params, -EINVAL);

    if (params->api_version != SOL_HTTP_PARAM_API_VERSION) {
        SOL_ERR("API version mistmatch; expected %u, got %u",
            SOL_HTTP_PARAM_API_VERSION, params->api_version);
        return false;
    }

    if (!params->arena) {
       params->arena = sol_arena_new();
       SOL_NULL_CHECK(params->arena, false);
    }

   if (value.type == SOL_HTTP_PARAM_QUERY_PARAM || ... SOL_HTTP_PARAM_COOKIE .. SOL_HTTP_PARAM_HEADER ... SOL_HTTP_PARAM_POST_FIELD) {
      int r;
      r = sol_arena_slice_dup(params->arena, &value.key_value.key, value.key_value.key);
      SOL_INT_CHECK(r, < 0, false);
      r = sol_arena_slice_dup(params->arena, &value.key_value.value, value.key_value.value);
      SOL_INT_CHECK(r, < 0, false);
   } else if (value.type == SOL_HTTP_PARAM_AUTH_BASIC) {
      int r;
      r = sol_arena_slice_dup(params->arena, &value.auth.user, value.auth.user);
      SOL_INT_CHECK(r, < 0, false);
      r = sol_arena_slice_dup(params->arena, &value.auth.password, value.auth.password);
      SOL_INT_CHECK(r, < 0, false);
   } else if (value.type == SOL_HTTP_PARAM_POST_DATA) {
      int r;
      r = sol_arena_slice_dup(params->arena, &value.data.value, value.data.value);
      SOL_INT_CHECK(r, < 0, false);  
   }

    ptr = sol_vector_append(&params->params);
    if (!ptr) {
        SOL_WRN("Could not append option to parameter vector");
        return false;
    }

    memcpy(ptr, &value, sizeof(value));
    return true;
}

SOL_API void
sol_http_param_free(struct sol_http_param *params)
{
    SOL_NULL_CHECK(params);

    if (params->api_version != SOL_HTTP_PARAM_API_VERSION) {
        SOL_ERR("API version mistmatch; expected %u, got %u",
            SOL_HTTP_PARAM_API_VERSION, params->api_version);
        return;
    }
    sol_vector_clear(&params->params);
    if (params->arena) { // <--- addition ----------------
       sol_arena_del(params->arena);
       params->arena = NULL;
    }
}